File: stacked-web/lib/utils/label-id-order.ts
/** Ordem de etiquetas no card: última tocada/aplicada primeiro (MRU). */

/** Liga → início; desliga → remove. Demais mantêm ordem relativa. */
export function toggleLabelIdOrder(ids: string[], id: string): string[] {
  if (ids.includes(id)) return ids.filter((x) => x !== id);
  return [id, ...ids.filter((x) => x !== id)];
}

/** Resolve catálogo na ordem dos ids. */
export function resolveLabelsByIds<T extends { id: string }>(
  ids: string[],
  catalog: T[],
): T[] {
  return ids
    .map((id) => catalog.find((l) => l.id === id))
    .filter((l): l is T => Boolean(l));
}

File: stacked-ios/Stacked/Services/LabelIdOrder.swift
import Foundation

/// Ordem de etiquetas no card: última tocada/aplicada primeiro (MRU).
enum LabelIdOrder {
  /// Liga → vai para o início; desliga → remove. Demais mantêm a ordem relativa.
  static func toggle(_ ids: [String], id: String) -> [String] {
    if ids.contains(id) {
      return ids.filter { $0 != id }
    }
    return [id] + ids.filter { $0 != id }
  }

  /// Resolve catálogo na ordem dos ids (não na ordem do catálogo).
  static func resolve<T>(_ ids: [String], from catalog: [T], id: (T) -> String) -> [T] {
    ids.compactMap { wanted in catalog.first(where: { id($0) == wanted }) }
  }
}
